### PR TITLE
LPS-66855

### PR DESCRIPTION
**Pull request by samuelkong**
Can you review my changes. Thanks.

### CODE REVIEW

**Comment by samuelkong** [comment]
ci:retest

File: portal-impl/src/com/liferay/portal/util/HtmlImpl.java [hunk 1]
```diff
@@ -109,15 +109,86 @@ public String escape(String text) {
...
+		if (sb == null) {
+			return text;
+		}
+
+		if (lastReplacementIndex < text.length()) {
+			sb.append(text.substring(lastReplacementIndex));
+		}
+
+		return sb.toString();
 	}
```
Comment (topolik) [inline portal-impl/src/com/liferay/portal/util/HtmlImpl.java]: Please see https://github.com/brianchandotcom/liferay-portal/commit/89e2e7bebbd2214caa600094e79570f20b7821e0

Should I update it the same way or leave it for Brian? :)

**Comment by samuelkong** [comment]
I thought about changing it, but ultimately decided on being lazy.

I went ahead and made the change and updated the branch / this pull request.

**Comment by topolik** [comment]
Thanks Sam. 

The hole goes deeper and we should re-implement HtmlImpl for this to be correct. Working with just one char breaks UTF-8, we need to work with code points.

----
Some thoughts:

What about surrogate pairs for the code points 0x10000-0x10FFF? I see HtmlImpl doesn't work with 4-bytes chars at all, but they are still valid for UTF-8.

And we need to fix this (control chars) also for CSS or URLs, they are also part of generated HTML. Basically everything that's printed into HTML. But we cannot cover everything because HtmlUtil is not used everywhere, for example article content :( I guess this should rather go to JspWriter or PrintWriter / ServletOutputStream, which we cannot influence in app servers :(

And for escapeAttribute we only escape chars with codepoint < 256! We missed for example \u2028, \u2029 or \uFEFF which are treated as white-space and text with these chars escapes from non-quoted attribute -> XSS vector.

And last, but not least, escape method is escaping characters that are there only because some browser bugs. 0x00bb, 0x2013 and 0x2014 are valid and not harmful.
----
So, in my opinion, every escape* method should go through a "filter" to remove control and invalid XML chars. For this the "filter" must work with Unicode `int` code points not Java `char`s.

Then, for example, we can implement escapeAttribute/CSS/JS/URL(int codePoint) and encode either ONLY harmful chars or everything for that specific context.

**Comment by samuelkong** [comment]
I totally agree. Long term code is to replace our implementation with OWASP Java Encoder. https://www.owasp.org/index.php/OWASP_Java_Encoder_Project This will take care of checking for surrogate pair issue. I need to find out why we're escaping some of the characters that are not necessary (thanks for the pointer about the browser bugs. I'll look into this). For now, I'm just trying to get past the Veracode process.

It's actually safe to have control characters in CSS and JS. However, I think you're right about URLs.

For escapeAttribute, I have a pending pull request with the docs team to specify that the escapeAttribute should only be used for quoted attributes. If we only allow quoted attributes, then we don't have to worry about  \u2028, \u2029 or \uFEFF.

**Comment by topolik** [comment]
Pull request submitted at: https://github.com/samuelkong/liferay-portal/pull/769

View just my changes: https://github.com/topolik/liferay-portal/compare/samuelkong:5e9d75d6fa...2033bd8f17
View total diff: https://github.com/topolik/liferay-portal/compare/a6a8f0d652...2033bd8f17

**Pull request closed without merging**